File: snippets/cpp/rq/segt.cpp
struct SegTree {
    int n; vi t;
    SegTree(int n) : n(n), t(4 * n) {}
    SegTree(vi &a) : n(sza(a)), t(4 * n) { build(a, 0, 0, n - 1); }
    void build(vi &a, int v, int l, int r) {
        if (l == r) { t[v] = a[l]; return; }
        int m = (l + r) / 2;
        build(a, 2*v+1, l, m); build(a, 2*v+2, m+1, r);
        t[v] = t[2*v+1] + t[2*v+2];
    }
    void upd(int i, int v, int x, int l, int r) {
        if (l == r) { t[x] = v; return; }
        int m = (l + r) / 2;
        if (i <= m) upd(i, v, 2*x+1, l, m);
        else upd(i, v, 2*x+2, m+1, r);
        t[x] = t[2*x+1] + t[2*x+2];
    }
    int qry(int ql, int qr, int x, int l, int r) {
        if (ql > r || qr < l) return 0;
        if (ql <= l && r <= qr) return t[x];
        int m = (l + r) / 2;
        return qry(ql, qr, 2*x+1, l, m) + qry(ql, qr, 2*x+2, m+1, r);
    }
    void upd(int i, int v) { upd(i, v, 0, 0, n - 1); }
    int qry(int l, int r) { return qry(l, r, 0, 0, n - 1); }
};

File: snippets/rust/rq/segt.rs
// Point-update, range-sum segment tree. Indices are 0..n.
struct SegTree {
    n: usize,
    t: Vec<i64>,
}
impl SegTree {
    fn new(n: usize) -> Self { Self { n, t: vec![0; 4 * n.max(1)] } }
    fn from(a: &[i64]) -> Self {
        let n = a.len();
        let mut s = Self::new(n);
        if n > 0 { s.build(a, 1, 0, n - 1); }
        s
    }
    fn build(&mut self, a: &[i64], v: usize, l: usize, r: usize) {
        if l == r { self.t[v] = a[l]; return; }
        let m = (l + r) / 2;
        self.build(a, 2 * v, l, m);
        self.build(a, 2 * v + 1, m + 1, r);
        self.t[v] = self.t[2 * v] + self.t[2 * v + 1];
    }
    fn update(&mut self, i: usize, val: i64) {
        self.upd(i, val, 1, 0, self.n - 1);
    }
    fn upd(&mut self, i: usize, val: i64, v: usize, l: usize, r: usize) {
        if l == r { self.t[v] = val; return; }
        let m = (l + r) / 2;
        if i <= m { self.upd(i, val, 2 * v, l, m); }
        else { self.upd(i, val, 2 * v + 1, m + 1, r); }
        self.t[v] = self.t[2 * v] + self.t[2 * v + 1];
    }
    fn query(&self, l: usize, r: usize) -> i64 {
        self.qry(l, r, 1, 0, self.n - 1)
    }
    fn qry(&self, ql: usize, qr: usize, v: usize, l: usize, r: usize) -> i64 {
        if qr < l || r < ql { return 0; }
        if ql <= l && r <= qr { return self.t[v]; }
        let m = (l + r) / 2;
        self.qry(ql, qr, 2 * v, l, m) + self.qry(ql, qr, 2 * v + 1, m + 1, r)
    }
}
